Src/FluentAssertions/Execution/AssertionChain.cs

```diff
@@ -0,0 +1,355 @@
+﻿using System;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using FluentAssertions.Common;
+
+namespace FluentAssertions.Execution;
+
+public sealed class AssertionChain
+{
+    private readonly Func<AssertionScope> getCurrentScope;
+    private readonly ContextDataDictionary contextData = new();
+    private string fallbackIdentifier = "object";
+    private Func<string> getCallerIdentifier;
+    private Func<string> reason;
+    private bool? succeeded;
+
+    // We need to keep track of this because we don't want the second successful assertion hide the first unsuccessful assertion
+    private Func<string> expectation;
+    private string callerPostfix = string.Empty;
+
+    private static readonly AsyncLocal<AssertionChain> Instance = new();
+
+    /// <summary>
+    /// Ensures that the next call to <see cref="GetOrCreate"/> will reuse the current instance.
+    /// </summary>
+    public void ReuseOnce()
+    {
+        Instance.Value = this;
+    }
+
+    /// <summary>
+    /// Indicates whether the previous assertion in the chain was successful.
+    /// </summary>
+    /// <remarks>
+    /// This property is used internally to determine if subsequent assertions
+    /// should be evaluated based on the result of the previous assertion.
+    /// </remarks>
+    internal bool PreviousAssertionSucceeded { get; private set; } = true;
+
+    /// <summary>
+    /// Indicates whether the caller identifier has been manually overridden.
+    /// </summary>
+    /// <remarks>
+    /// This property is used to track if the caller identifier has been customized using the
+    /// <see cref="OverrideCallerIdentifier"/> method or similar methods that modify the identifier.
+    /// </remarks>
+    public bool HasOverriddenCallerIdentifier { get; private set; }
+
+    /// <summary>
+    /// Either starts a new assertion chain, or, when <see cref="ReuseOnce"/> was called, for once, will return
+    /// an existing instance.
+    /// </summary>
+    public static AssertionChain GetOrCreate()
+    {
+        if (Instance.Value != null)
+        {
+            AssertionChain assertionChain = Instance.Value;
+            Instance.Value = null;
+            return assertionChain;
+        }
+
+        return new AssertionChain(() => AssertionScope.Current,
+            () => FluentAssertions.CallerIdentifier.DetermineCallerIdentity());
+    }
+
+    private AssertionChain(Func<AssertionScope> getCurrentScope, Func<string> getCallerIdentifier)
```

Comment:
The one and only constructor is private?
This makes writing some tests very complicated, at least for extensions of FA.

Reply:
You're supposed to use `GetOrCreate`. Can you explain why you would need to access this constructor directly?

Reply:
Currently I do not have an explicit use case.
But for writing tests explicit with or for `AssertionChain` I must not use a static instance.
This is an experience of writing public libraries. Everything which I build initially private or internal will get some day public...

Reply:
This is absolutely true. However, `AssertionChain.GetOrCreate` always returns a new instance, unless you use `ReuseOnce`, but that only works once, and travels over the async control flow.